plugins/woocommerce/src/Internal/AddressProvider/AbstractAutomatticAddressProvider.php

```diff
@@ -0,0 +1,165 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\Internal\AddressProvider;
+
+use Automattic\WooCommerce\StoreApi\Utilities\JsonWebToken;
+use Automattic\Jetpack\Constants;
+use WC_Address_Provider;
+
+/**
+ * Abstract Automattic address provider is an abstract implementation of the WC_Address_Provider that is meant to be used by Automattic services to get support for address autocomplete and maps with minimal code maintenance.
+ *
+ * @since 10.1.0
+ * @package WooCommerce
+ */
+abstract class AbstractAutomatticAddressProvider extends WC_Address_Provider {
+
+	/**
+	 * The JWT for the address service.
+	 *
+	 * @var string
+	 */
+	private $jwt = null;
+
+	/**
+	 * Loads up the JWT for the address service and saves it to transient.
+	 */
+	public function __construct() {
+		add_filter( 'pre_update_option_woocommerce_address_autocomplete_enabled', array( $this, 'refresh_cache' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'load_scripts' ) );
+	}
+
+	/**
+	 * Get the JWT for the address service, a service should implement an A8C hosted API or some mechanism to get a JWT, this will be passed to frontend code to be used in the address autocomplete and maps.
+	 *
+	 * This method shouldn't implement any caching, it should only fetch the token or throw an exception, if you must handle caching, consider also overriding get_jwt.
+	 *
+	 * @return string The JWT for the address service.
+	 */
+	abstract public function get_address_service_jwt();
+
+	/**
+	 * Loads up a JWT from cache or from the implementor side.
+	 *
+	 * @return string|null The JWT for the address service.
+	 */
+	public function load_jwt() {
+
+		// If we already have a loaded, valid token, we return early.
+		if ( $this->jwt && JsonWebToken::shallow_validate( $this->jwt ) ) {
+			return $this->jwt;
+		}
+
+		$transient_key = $this->id . 'address_autocomplete_jwt';
+		$cached_jwt    = get_transient( $transient_key );
+		// If we have a cached, valid token, we load it to class and return early.
+		if ( $cached_jwt && JsonWebToken::shallow_validate( $cached_jwt ) && 'local' !== wp_get_environment_type() ) {
+			$this->jwt = $cached_jwt;
+			return $this->jwt;
+		}
+
+		// Otherwise, we fetch a fresh token.
+		try {
+			$fresh_jwt = $this->get_address_service_jwt();
+			if ( $fresh_jwt && JsonWebToken::shallow_validate( $fresh_jwt ) ) {
+				$this->jwt = $fresh_jwt;
+				set_transient( $transient_key, $fresh_jwt, DAY_IN_SECONDS );
```

Comment:
A token is valid for one day, I think we can maybe change this so the token is value for its expiry date.